# Проект модуля. Нейросеть для автодополнения текстов

Примечания перед началом:
* как минимум один csv-файл потребуется, файл c sentiment140
* поместить свой или достатый из zip-архива файл c вышеупомянутым набором данных в папку дата, переименовать его в 'raw_dataset.csv'
* если достать остальные из архивов, то можно сэкономить немного времени и не проходить ветку за else
* использовал python 3.14.3

In [1]:
import sys
sys.path.append('src')  # добавляем путь к директории src, чтобы Python смог-таки находить модули там
import src.data_utils as du
import src.combine as cmb
import src.transformer as tf



if(du.check_files_existence(du.csv_files_to_check)):
    print("Все csv файлы существуют, уже легче")
else:
    print("Один или несколько csv файлов отсутствуют")
    du.extract_texts_from_to(du.source_csv, du.tweets_csv)
    du.preprocess_data(du.tweets_csv, du.preprocess_csv)


cmb.run_lstm_experiment(du.preprocess_csv)
tf.run_transformer_run('data/val_samples.csv')
tf.gpt2_generate_sequence('data/test_samples.csv')

Все csv файлы существуют, уже легче
Данные прочитаны
Выполнено разбиение на выборки
Выборка успешно сохранена в файл data/train_samples.csv
Выборка успешно сохранена в файл data/val_samples.csv
Выборка успешно сохранена в файл data/test_samples.csv
Выборки сохранены


Токенайзер загружен
Создание наборов данных...
Наборы данных и загрузчики созданы
Модель создана
Начинаем процесс обучения:


Epoch 1/10:   0%|          | 0/25 [00:00<?, ?it/s]

Epoch 1: Average Training Loss=4.7186
Epoch 1: Validation Loss=3.8101
Расчёт метрик ROUGE...
Расчёт метрик ROUGE завершён
Сохранение всякого 18+...
Создана промежуточная точка восстановления модели ./models/model_epoch_1.pth


Epoch 2/10:   0%|          | 0/25 [00:00<?, ?it/s]

Epoch 2: Average Training Loss=2.7982
Epoch 2: Validation Loss=3.6349
Расчёт метрик ROUGE...
Расчёт метрик ROUGE завершён
Сохранение всякого 18+...
Создана промежуточная точка восстановления модели ./models/model_epoch_2.pth


Epoch 3/10:   0%|          | 0/25 [00:00<?, ?it/s]

Epoch 3: Average Training Loss=2.6416
Epoch 3: Validation Loss=3.6168
Расчёт метрик ROUGE...
Расчёт метрик ROUGE завершён
Сохранение всякого 18+...
Создана промежуточная точка восстановления модели ./models/model_epoch_3.pth


Epoch 4/10:   0%|          | 0/25 [00:00<?, ?it/s]

Epoch 4: Average Training Loss=2.5961
Epoch 4: Validation Loss=3.6372
Расчёт метрик ROUGE...
Расчёт метрик ROUGE завершён
Сохранение всякого 18+...
Создана промежуточная точка восстановления модели ./models/model_epoch_4.pth


Epoch 5/10:   0%|          | 0/25 [00:00<?, ?it/s]

Epoch 5: Average Training Loss=2.5613
Epoch 5: Validation Loss=3.6152
Расчёт метрик ROUGE...
Расчёт метрик ROUGE завершён
Сохранение всякого 18+...
Создана промежуточная точка восстановления модели ./models/model_epoch_5.pth


Epoch 6/10:   0%|          | 0/25 [00:00<?, ?it/s]

Epoch 6: Average Training Loss=2.5170
Epoch 6: Validation Loss=3.6043
Расчёт метрик ROUGE...
Расчёт метрик ROUGE завершён
Сохранение всякого 18+...
Создана промежуточная точка восстановления модели ./models/model_epoch_6.pth


Epoch 7/10:   0%|          | 0/25 [00:00<?, ?it/s]

Epoch 7: Average Training Loss=2.4589
Epoch 7: Validation Loss=3.6068
Расчёт метрик ROUGE...
Расчёт метрик ROUGE завершён
Сохранение всякого 18+...
Создана промежуточная точка восстановления модели ./models/model_epoch_7.pth


Epoch 8/10:   0%|          | 0/25 [00:00<?, ?it/s]

Epoch 8: Average Training Loss=2.3889
Epoch 8: Validation Loss=3.6228
Расчёт метрик ROUGE...
Расчёт метрик ROUGE завершён
Сохранение всякого 18+...
Создана промежуточная точка восстановления модели ./models/model_epoch_8.pth


Epoch 9/10:   0%|          | 0/25 [00:00<?, ?it/s]

Epoch 9: Average Training Loss=2.3128
Epoch 9: Validation Loss=3.6582
Расчёт метрик ROUGE...
Расчёт метрик ROUGE завершён
Сохранение всякого 18+...
Создана промежуточная точка восстановления модели ./models/model_epoch_9.pth


Epoch 10/10:   0%|          | 0/25 [00:00<?, ?it/s]

Epoch 10: Average Training Loss=2.2326
Epoch 10: Validation Loss=3.6641
Расчёт метрик ROUGE...
Расчёт метрик ROUGE завершён
Сохранение всякого 18+...
Создана промежуточная точка восстановления модели ./models/model_epoch_10.pth
Обучение завершено
*************************


--- Последние записи истории обучения ---
Train_loss:	2.2325772190093995
Val_loss:	3.6641167998313904
Rouge:	{'rouge1': np.float64(0.0), 'rouge2': np.float64(0.0), 'rougeL': np.float64(0.0), 'rougeLsum': np.float64(0.0)}
*************************

Тексты тестового набора:
0                 good god they ruined my belly button
1                                       c mon sean man
2    i think maybe you should get a couple more hou...
3    i miss my ps3 it s out of commission wutcha pl...
4    hey did u talk to mom r they home yet i hope t...
Name: text, dtype: str

Tokens: [101, 2204, 2643, 2027, 9868, 2026, 7579, 6462, 102], [CLS] good god they ruined my belly button [SEP]
First group: [101, 2204, 2643, 2027, 9868,

Loading weights:   0%|          | 0/76 [00:00<?, ?it/s]

Processing Batches:   0%|          | 0/4 [00:00<?, ?it/s]



--------------------------------------------------
:жпт2 метрики руж:
{'rouge1': np.float64(0.05353110207608457), 'rouge2': np.float64(0.003857142857142858), 'rougeL': np.float64(0.05231531047959234), 'rougeLsum': np.float64(0.0521529462311667)}


Loading weights:   0%|          | 0/76 [00:00<?, ?it/s]



--------------------------------------------------
:жпт2 пример #1:
Исходный текст: good god they ruined my belly button
Входная последовательность: good god they ruined my
Полный текст с продолжением: good god they ruined my life.※
I think it※s wrong to say that I do. The most
--------------------------------------------------


--------------------------------------------------
:жпт2 пример #2:
Исходный текст: c mon sean man
Входная последовательность: c mon se
Полный текст с продолжением: c mon se lange.
--------------------------------------------------


--------------------------------------------------
:жпт2 пример #3:
Исходный текст: i think maybe you should get a couple more hours of sleep hon how productive can you be right now if ur dog tired i worry
Входная последовательность: i think maybe you should get a couple more hours of sleep hon how productive can you be right
Полный текст с продолжением: i think maybe you should get a couple more hours of sleep hon how productiv

*Выводы по модели LSTM после обучения на 800 примерах за 10 эпох:*
* Lstm-модель дополняет предложения, значит худо-бедно обучаема
* Метрики ROUGE нулевые, значит модель выдаёт что-то несуразное
* Стиль генерируемых последовательностей соответствует стилю исходного набора - птичий
* Для большей суразности:
  * попробовать увеличить объём данных и время обучения
  * изменить способ токенизации
  * изменить характеристики модели
  * изменить параметры обучения


*Выводы по использованию готовой gpt-2:*
* {'rouge1': np.float64(0.05353110207608457), 'rouge2': np.float64(0.003857142857142858), 'rougeL': np.float64(0.05231531047959234), 'rougeLsum': np.float64(0.0521529462311667)} - низкое значение метрик, значит генерируемый текст сильно отличается от образца.
* Стиль генерируемых продолжений отличается от птичьего - это уже стиль человека с немного развитым речевым аппаратом, возможно это причина промахов, возможно опять токенизация

**Общий вывод:**
На данный момент ни одна из моделей не годится для автодополнения, если не стоит цели максимально вывести из себя пользователя
